### ISSUE TITLE
Lerna build

### ISSUE THREAD

**Issue by raulk** [ConsenSys/ethql#107]
## Motivation

We want to publish each plugin independently on NPM, so that users can compose only the plugins they require in their deployment. Independent versioning of plugins may also be desirable.

## Requirements

- [x] Configure the build to use [lerna](https://github.com/lerna/lerna).
- [x] Each plugin should be enclosed in a separate package, with the exception of `core`, which should be part of the engine.
- [x] Resulting Node packages:
  * ethql => foundation + core plugin
  * ethql-erc20
  * ethql-ens (see #106)
- [x] Help us decide if we want independent module versioning, or not. Argument for and against.
- [x] This issue should be resolved after #106, so that we have at least 3 packages.
- [x] **do not publish anything on NPM just yet, we will do that from the ConsenSys organisation**.

## Definition of done

- [x]  PR accepted and merged, including all changes to `package.json`, directory layout, etc. to implement the above requirements.
- [ ] Updated README and wiki pages: particularly "how to build" instructions, and "usage/install" sections.
- [ ] Good discussion with a decision regarding independent versioning or not.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 250.0 DAI (250.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ConsenSys/ethql/107/1287).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ConsenSys/ethql/107/1287).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $22,946.04 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 2 months, 4 weeks from now.
Please review their action plans below:


**1) [mul1sh](https://gitcoin.co/profile/mul1sh) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?mutate_worker_action=approve&worker=mul1sh) | [reject worker](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?mutate_worker_action=reject&worker=mul1sh))_.**

I have published a few node modules/plugins to npm, so I believe i can tackle this issue.

My approach will simply be to break down the ethql project into the required plugins using lerna which i'm also quite conversant with.

I should finish this in less than 4 days, thanks

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ConsenSys/ethql/107/1287).

**Comment by ceresstation** [comment]
Hey @mul1sh I see you already have quite a few large bounties in progress, are you sure you have time to take this on as well?

**Comment by mul1sh** [comment]
@ceresstation yep I believe I can handle this one as well because i'm almost done with 2 my other bounties and the other one is WIP but isn't that urgent because of major blocker we discovered in its implementation. So i believe I can handle this issue as I finish up and submit the other bounties, and submit it by Tuesday next week of course with daily updates, 😃  thanks.

**Comment by gitcoinbot** [comment]
@mul1sh Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=1) | [3 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=3) | [5 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=5) | [10 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=10) | [100 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=100)

**Comment by mul1sh** [comment]
uuhhmmm.....already made my PR @gitcoinbot 😂 😂 , so yes i'm still working on this issue

**Comment by mul1sh** [comment]
@raulk btw has issue #106  been solved so that I continue working on this ticket?? thanks 🙂

**Comment by mul1sh** [comment]
@raulk @c-castillo any help with the above 👆 query? or should I continue working and stop waiting for issue 106 to be closed?

**Comment by gitcoinbot** [comment]
@mul1sh Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=1) | [3 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=3) | [5 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=5) | [10 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=10) | [100 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=100)

**Comment by gitcoinbot** [comment]
@mul1sh Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=1) | [3 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=3) | [5 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=5) | [10 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=10) | [100 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=100)

**Comment by mul1sh** [comment]
@gitcoinbot yeah i'm working on the issue 😃

**Comment by raulk** [comment]
@mul1sh issue #106 has not been solved yet, as conveyed by the status "open". There is a WIP pull request in progress here: #111. You can track that for updates.

You can update your WIP PR to make sure we're not introducing a package-lock file, and that we're integrating with yarn.

**Comment by raulk** [comment]
@mul1sh Could you provide an estimate and a plan to move forward? I am not satisfied with the WIP PR #110, which only added basic boilerplate and didn't even restructure the tree in packages. That PR did not honor our usage of yarn, and instead attempted to introduce a `package-lock.json` as a result of plain NPM usage.

Please advise if you're still interested in pursuing this issue; if so, we'll need to see substantial progress in the next day, otherwise we'll release the bounty back to the community.

**Comment by mul1sh** [comment]
@raulk sincere apologies for the delay, i truly wanted to finish this issue this week but things were abit chaotic for me.

Nonetheless i'm still very much interested in this issue and I plan to work on it from Friday and over the weekend and deliver it on Monday. So please expect another PR from me in less than 24 hours with more substantial changes.

Thanks.

**Comment by mul1sh** [comment]
@raulk i've added the package tree, i'm now separating each of the dependencies of each package into a seperate `package.json` file. So i'm course to finishing this tomorrow.

2 questions though, 

1) For the ethql package, what constitutes the foundation + core plugin?
2) Is the ens module ready for integration and if so, what are the respective files and folders to use?

Thanks for your patience, I highly appreciate it.

**Comment by mul1sh** [comment]
@StevenJNPearce  i'm working on this issue and it requires #106 to be working, any idea when it would be ready so that I incorporate it into my PR? Thanks 🙂

**Comment by raulk** [comment]
@mul1sh I believe I provided answers to those questions in #113. Note that getting #106 merged is not an impediment to nail the Lerna structure down, which is what this bounty is about. Once we were happy with the Lerna structure, converting the ENS plugin to it would've been trivial.

Unfortunately we haven't seen progress in the right direction, so we'll be putting this bounty up for grabs again in Gitcoin. Thanks!

**Comment by mul1sh** [comment]
@raulk  ok well noted, apologies again for the delay, i'll stop work so that someone else can pick it up.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 5 months ago.
Please review their action plans below:


**1) [evgeniuz](https://gitcoin.co/profile/evgeniuz) has been approved to start work.**

I would like to make this migration, was trying out lerna recently :) I will also check what versioning scheme seems to me more relevant for this case and provide my thoughts.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ConsenSys/ethql/107/1287).

**Comment by evgeniuz** [comment]
Thanks, already started working on it. Will create WIP PR once I resolve issue with TypeScript typing between packages (during my quick tests web3 types were not resolved correctly). I will try to separate it in @types package, so that it can be removed when patches are merged in DefinitelyTyped.

**Comment by evgeniuz** [comment]
As for independent versioning or not: my suggestion is go with independent at least up to version 1.0.0. Semantic versioning suggests that no API should break when minor/patch version are updated, so until you introduce breaking changes it's easier to maintain common versioning, than to track them separately (that's why it's default in lerna).

After version 1.0.0 I think it's still makes sense to keep common versioning: since packages are mostly plugins for ethql itself, it's unlikely that some other software will rely on a plugin *without* ethql, so they probably should be upgraded together.

There are two scenarios I can think of where independent versioning is better:

- ethql plugins will receive lots of updates, while ethql core remains same for a long time;
- some other packages are introduced that can be used without ethql (e. g. some utility package useful in other context)

In this case you can switch to independent versioning by adding a single parameter to `lerna.json`, which can be done at any time.

**Comment by gitcoinbot** [comment]
@evgeniuz Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=1) | [3 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=3) | [5 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=5) | [10 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=10) | [100 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=100)

**Comment by gitcoinbot** [comment]
@evgeniuz Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=1) | [3 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=3) | [5 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=5) | [10 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=10) | [100 days](https://gitcoin.co/issue/ConsenSys/ethql/107/1287?snooze=100)

**Comment by ceresstation** [comment]
Hey @raulk have you had a chance to review @evgeniuz's work?

**Comment by kshinn** [comment]
@evgeniuz Sorry for the delay. Are you still interested in finishing this bounty? I've merged this branch into upstream consensus branch and I believe the only thing left to do is to update the README. At that point, I think we can consider this done and merged.

**Comment by evgeniuz** [comment]
@kshinn Sure, will update README in a day or so.

**Comment by kshinn** [comment]
Great. If you can do that in a separate PR, I'm going to merge the original PR into master since it is blocking other work.

**Comment by evgeniuz** [comment]
@kshinn just checked, README in `lerna` branch is already updated: those three commands (`install`, `bootstrap`, `dev`) is enough to start development. Lerna itself doesn't need any special moves for development (unless new module is added, but this is not a topic for README). Not sure if I need to add anything there.

**Comment by kshinn** [comment]
Thanks for checking. Considering this work finished.

**Comment by kshinn** [comment]
@ceresstation Are we able to pay this bounty?

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 250.0 DAI (250.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@evgeniuz](https://gitcoin.co/profile/evgeniuz)



@ceresstation please take a look at the submitted work:
* [PR](https://github.com/ConsenSys/ethql/pull/117) by @evgeniuz

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ConsenSys/ethql/107/1287)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ConsenSys/ethql/107/1287).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $99,133.95 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__The funding of 250.0 DAI (250.0 USD @ $1.0/DAI)  attached to this issue has been approved & issued to @evgeniuz.__  

 
* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ConsenSys/ethql/107/1287)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $83,736.07 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
